### PR TITLE
Impl generics

### PR DESCRIPTION
**Pull request by flodiebold**
This handles type parameters on impls when typing method calls.

~One remaining problem is that the autoderefs aren't applied during the unification of the method receiver type with the actual receiver type, which means that the type parameters are only correctly inferred if no autoderefs happened.~

### CODE REVIEW

File: crates/ra_hir/src/path.rs [hunk 1]
```diff
@@ -129,7 +129,7 @@ impl Path {
 }
 
 impl GenericArgs {
-    fn from_ast(node: &ast::TypeArgList) -> Option<GenericArgs> {
+    pub fn from_ast(node: &ast::TypeArgList) -> Option<GenericArgs> {
```
Comment (matklad) [inline crates/ra_hir/src/path.rs]: should be `pub(crate)`? (there's assist to swtitc to pub(crate) from pub)

**Comment by matklad** [comment]
r=me with visibility fixed!

File: crates/ra_hir/src/path.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/path.rs]: I used the assist ;)

**Comment by flodiebold** [comment]
bors r=matklad

**Pull request merged**